phpunit/class-wp-get-block-css-selectors-test.php

```diff
@@ -0,0 +1,319 @@
+<?php
+/**
+ * Tests_Get_Block_CSS_Selector class
+ *
+ * @package WordPress
+ */
+
+class WP_Get_Block_CSS_Selector_Test extends WP_UnitTestCase {
+	private $test_block_name;
+
+	public function set_up() {
+		parent::set_up();
+		$this->test_block_name = null;
+	}
+
+	public function tear_down() {
+		unregister_block_type( $this->test_block_name );
+		$this->test_block_name = null;
+		parent::tear_down();
+	}
+
+	private function register_test_block( $name, $selectors = null, $supports = null ) {
+		$this->test_block_name = $name;
+
+		return register_block_type(
+			$this->test_block_name,
+			array(
+				'api_version' => 2,
+				'attributes'  => array(),
+				'selectors'   => $selectors,
+				'supports'    => $supports,
+			)
+		);
+	}
+
+	public function test_get_root_selector_via_selectors_api() {
+		$block_type = self::register_test_block(
+			'test/block-with-selectors',
+			array( 'root' => '.wp-custom-block-class' )
+		);
+
+		$selector = wp_get_block_css_selector( $block_type );
+		$this->assertEquals( '.wp-custom-block-class', $selector );
+	}
+
+	public function test_get_root_selector_via_experimental_property() {
+		$block_type = self::register_test_block(
+			'test/block-without-selectors',
+			null,
+			array( '__experimentalSelector' => '.experimental-selector' )
+		);
+
+		$selector = wp_get_block_css_selector( $block_type );
+		$this->assertEquals( '.experimental-selector', $selector );
+	}
+
+	public function test_default_root_selector_generation() {
+		$block_type = self::register_test_block(
+			'test/without-selectors-or-supports',
+			null,
+			null
+		);
+
+		$selector = wp_get_block_css_selector( $block_type );
+		$this->assertEquals( '.wp-block-test-without-selectors-or-supports', $selector );
+	}
+
+	public function test_get_duotone_selector_via_selectors_api() {
+		$block_type = self::register_test_block(
+			'test/duotone-selector',
+			array(
+				'color' => array( 'duotone' => '.duotone-selector' ),
+			),
+			null
+		);
+
+		$selector = wp_get_block_css_selector( $block_type, 'duotone' );
+		$this->assertEquals( '.duotone-selector', $selector );
+	}
+
+	public function test_get_duotone_selector_via_experimental_property() {
+		$block_type = self::register_test_block(
+			'test/experimental-duotone-selector',
+			null,
+			array(
+				'color' => array(
+					'__experimentalDuotone' => '.experimental-duotone',
+				),
+			)
+		);
+
+		$selector = wp_get_block_css_selector( $block_type, 'duotone' );
+		$this->assertEquals( '.experimental-duotone', $selector );
+	}
+
+	public function test_no_duotone_selector_set() {
+		$block_type = self::register_test_block(
+			'test/null-duotone-selector',
+			null,
+			null
+		);
+
+		$selector = wp_get_block_css_selector( $block_type, 'duotone' );
```

Comment:
The API signature says that it returns the CSS selector (string) or `null` if none is available. Why would this return an empty string for `duotone`?

Reply:
Actually! Linking with this other conversation: https://github.com/WordPress/gutenberg/pull/46496/files#r1090917587

It sounds like the output of this method is always a CSS selector (string) except for those properties we don't want to fall back if there is no specific selector for them: at the moment, only `duotone`. Would that be true? If so, we may want to document the properties for which `null` is a valid return.

For the record, I don't have a strong opinion on whether using `null` or empty string for "no selector". My current struggle is to understand why we need both, as it seems only `duotone` can run into the use case of "no selector"?

Reply:
This was a typo on my part sorry. I believe the assertion should have checked for `null`. I fixed this in https://github.com/WordPress/gutenberg/pull/46496/commits/6669ac19b6daa53a0f702fd51f7c36516408ac4c.

I think the confusion here is that we have two slightly overlapping cases. 

1. Duotone: This feature can return `null` regardless of the `$fallback` param.
2. Everything else: Any other feature or subfeature can return `null` depending on whether we want to fallback to a broader selector.

I'd be happy to improve the documentation around this as we settle on an approach or confirm the current one. I'll expand on the fallbacks in my reply to https://github.com/WordPress/gutenberg/pull/46496#discussion_r1090917587.